File: CubeJava/CubeEntryHashSet.java
import java.util.List;

public class CubeEntryHashSet {
    public static int step = 0, size = 1;
    public static SinglyLinkedHashSet<List<TRFaces>> all = new SinglyLinkedHashSet<>();
    public static void main(String[] args) {
        all.add(TRFaces.ThreeFacesArray);
        SinglyLinkedHashSet.Entry<List<TRFaces>> n = all.head;
        while (n != null) {
            for (int i = 0; i < size; i++) {
                for (int j = 0; j < 6; j++) {
                    all.add(TRFaces.rotate(TRFaces.faceRotationIndices[2 + j % 2], TRFaces.faceRotationIndices[j],
                    n.element, TRFaces.rotation[j]));
                }
                n = n.after;
                if (n == null) break;
            }
            size = all.size() - size;
            System.out.printf("addStates: %d, step: %d\n", size, ++step);
        }
        System.out.printf("allSize: %d", all.size());
    }
} /*addStates: 6, step: 1
addStates: 28, step: 2
addStates: 131, step: 3
addStates: 611, step: 4
addStates: 2821, step: 5
addStates: 12909, step: 6
addStates: 58613, step: 7
addStates: 260276, step: 8
addStates: 1084359, step: 9
addStates: 3652195, step: 10
addStates: 6361743, step: 11
Exception in thread "main" java.lang.NullPointerException: Cannot read field "element" because "<local1>" is null
        at CubeEntryHashSet.main(CubeEntryHashSet.java:14)

addStates: 6361743, step: 11
addStates: 4660737, step: 12
allSize: 11022480
 */

File: CubePy/3FacesCubes/CubeEntryHashSet.py
class SinglyLinkedHashSet:
    class Entry:
        def __init__(self, element):
            self.element = element
            self.after = None

    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0
        self._set = set()

    def add(self, element):
        if element in self._set: return

        self._set.add(element)
        new_entry = self.Entry(element)

        if self.tail is None:
            self.head = self.tail = new_entry
        else:
            self.tail.after = new_entry
            self.tail = new_entry

        self.size += 1

from TRFaces import TRFaces
class CubeEntryHashSet:
    step = 0
    size = 1
    all = SinglyLinkedHashSet()

    @classmethod
    def main(cls):
        cls.all.add(TRFaces.ThreeFacesArray)
        n = cls.all.head
        
        while n is not None:
            current_size = cls.size
            cls.size = cls.all.size - cls.size

            for _ in range(current_size):
                if n is None: break

                for j in range(6):
                    rotated = TRFaces.rotate(
                        TRFaces.face_rotation_indices[2 + j % 2], TRFaces.face_rotation_indices[j],
                        n.element, TRFaces.rotation[j]
                    )
                    cls.all.add(rotated)
                n = n.after

            cls.step += 1
            print(f"addStates: {current_size}, step: {cls.step}")
            
        print(f"allSize: {cls.all.size}")

if __name__ == "__main__": CubeEntryHashSet.main()